router/usecase/pools/routable_cw_orderbook_pool.go

```diff
@@ -0,0 +1,257 @@
+		}
+		tick := r.OrderbookData.Ticks[tickIdx]
+
+		// Increment or decrement the current tick index depending on out order direction
+		switch directionOut {
+		case domain.BID:
+			tickIdx--
+		case domain.ASK:
+			tickIdx++
+		default:
+			return sdk.Coin{}, domain.OrderbookPoolInvalidDirectionError{Direction: directionIn}
+		}
+
+		// Calculate the price for the current tick
+		tickPrice, err := clmath.TickToPrice(tick.TickId)
+		if err != nil {
+			return sdk.Coin{}, err
+		}
+
+		// Output amount that should be filled given the current tick price
+		outputAmount := convertValue(amountInRemaining, tickPrice, directionOut)
+
+		// Tick liquidity for output side
+		outputTickLiquidity, err := getTickLiquidity(&tick.TickLiquidity, directionOut)
+		if err != nil {
+			return sdk.Coin{}, err
+		}
+
+		// How much of the order this tick can fill
+		outputFilled := getFillableAmount(outputTickLiquidity, outputAmount)
+
+		// How much of the input denom has been filled
+		inputFilled := convertValue(outputFilled, tickPrice, directionIn)
+
+		// Add the filled amount to the order total
+		amountOutTotal.AddMut(outputFilled)
+
+		// Subtract the filled amount from the remaining amount of tokens in
+		amountInRemaining.SubMut(inputFilled)
+	}
+
+	// Return total amount out
+	return sdk.Coin{Denom: r.TokenOutDenom, Amount: amountOutTotal.Dec().TruncateInt()}, nil
+}
+
+// GetTokenOutDenom implements RoutablePool.
+func (r *routableOrderbookPoolImpl) GetTokenOutDenom() string {
+	return r.TokenOutDenom
+}
+
+// String implements domain.RoutablePool.
+func (r *routableOrderbookPoolImpl) String() string {
+	return fmt.Sprintf("pool (%d), pool type (%d) Orderbook, pool denoms (%v), token out (%s)", r.ChainPool.PoolId, poolmanagertypes.CosmWasm, r.GetPoolDenoms(), r.TokenOutDenom)
+}
+
+// ChargeTakerFee implements sqsdomain.RoutablePool.
+// Charges the taker fee for the given token in and returns the token in after the fee has been charged.
+func (r *routableOrderbookPoolImpl) ChargeTakerFeeExactIn(tokenIn sdk.Coin) (tokenInAfterFee sdk.Coin) {
+	tokenInAfterTakerFee, _ := poolmanager.CalcTakerFeeExactIn(tokenIn, r.GetTakerFee())
+	return tokenInAfterTakerFee
+}
+
+// GetTakerFee implements domain.RoutablePool.
+func (r *routableOrderbookPoolImpl) GetTakerFee() math.LegacyDec {
+	return r.TakerFee
+}
+
+// SetTokenOutDenom implements domain.RoutablePool.
+func (r *routableOrderbookPoolImpl) SetTokenOutDenom(tokenOutDenom string) {
+	r.TokenOutDenom = tokenOutDenom
+}
+
+// CalcSpotPrice implements domain.RoutablePool.
+func (r *routableOrderbookPoolImpl) CalcSpotPrice(ctx context.Context, baseDenom string, quoteDenom string) (osmomath.BigDec, error) {
+	// Get the expected order direction
+	direction, err := r.GetDirection(quoteDenom, baseDenom)
+	if err != nil {
+		return osmomath.BigDec{}, err
+	}
+
+	tickIdx, err := r.GetStartTickIndex(direction.Opposite())
+	if err != nil {
+		return osmomath.BigDec{}, err
+	}
+
+	tick := r.OrderbookData.Ticks[tickIdx]
```

Comment:
This can be a separate function that is shared with ` CalculateTokenOutByTokenIn`

Reply:
Let's not over abstract it here because both `directionIn` and `directionOut` are still being used down the line. We can have multiple return and such, but I think that's an overkill.

Reply:
OK